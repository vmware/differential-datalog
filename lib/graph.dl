/* Functions and transformers for use in graph processing */

/* Compute strongly connected components of a directed graph:
 *
 * Type variables:
 * - `'E` - type that represents graph edge
 * - `'N` - type that represents graph node
 *
 * Arguments:
 * - `Edges` - relation that stores graph edges
 * - `from`  - extracts the source node of an edge
 * - `to`    - extracts the destination node of an edge
 *
 * Output:
 * - `Labels` - partitions a subset of graph nodes into strongly
 *    connected components represented by labeling these nodes
 *    with SCC ids.  An SCC id is id of the smallest node in the SCC.
 *    A node not covered by the partitioning belongs to an SCC
 *    with a single element (itself).
 */
extern transformer SCC(Edges:   relation['E],
                       from:    function(e: 'E): 'N,
                       to:      function(e: 'E): 'N)
    -> (SCCLabels: relation [('N, 'N)])

/* Compute connected components of a graph
 * (by propagating node IDs forward, retaining the minimum ID)
 *
 * Type variables:
 * - `'E`  - graph edge
 * - `'N`  - graph node id
 *
 * Arguments:
 * - `Edges`  - relation that stores graph edges
 * - `from`   - extracts the source node of an edge
 * - `to`     - extracts the destination node of an edge
 *
 * Output:
 * - `Labels` - Final labeling of node `n` with the smallest node
 *    id `l` such that `n` is reachable from `l`.
 */
extern transformer ConnectedComponents(Edges:   relation['E],
                                       from:    function(e: 'E): 'N,
                                       to:      function(e: 'E): 'N)
    -> (CCLabels: relation [('N, 'N)])

/* Optimized version of `ConnectedComponents` for graph with nodes of type
 * `bit<n>`, n <= 64.
 */
extern transformer ConnectedComponents64(Edges:   relation['E],
                                         from:    function(e: 'E): 'N,
                                         to:      function(e: 'E): 'N)
    -> (CCLabels: relation [('N, 'N)])

/* Compute a subset of bidirectional graph edges, i.e., all edges `(x,y)` such
 * that `(y,x)` is also an edge in the graph.
 *
 * This transformer is equivalent to the following DDlog rule, but can be more
 * memory efficient, as it only computes one arrangement of `Edges` instead of two
 * (note, however, that if your program already uses at least one of the
 * two arrangements elsewhere then this transformer will not improve on its
 * memory footprint).
 * ```
 * BiEdges(x,y) :- Edges(x,y), Edges(y,x).
 * ```
 *
 * WARNING: This transformer assumes that the `Edges` relation is distinct,
 * e.g., it is an input relation.  It will return INCORRECT result otherwise;
 * hence the `Unsafe` prefix.
 *
 * Type variables:
 * - `'E`  - graph edge
 * - `'N`  - graph node id
 *
 * Arguments:
 * - `Edges`  - relation that stores graph edges
 * - `from`   - extracts the source node of an edge
 * - `to`     - extracts the destination node of an edge
 *
 * Output:
 * - `BiEdges` - the subset of all bidirectional edges of the graph.
 *   Note: this subset will contain edges in both directions, i.e., for every
 *   `(x,y)` it also contains a `(y,x)`
 */
extern transformer UnsafeBidirectionalEdges(Edges:   relation['E],
                                            from:    function(e: 'E): 'N,
                                            to:      function(e: 'E): 'N)
    -> (BiEdges: relation [('N, 'N)])
