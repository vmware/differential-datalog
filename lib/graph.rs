/* Functions and transformers for use in graph processing */

use differential_dataflow::algorithms::graphs::propagate;
use differential_dataflow::algorithms::graphs::scc;
use differential_dataflow::collection::Collection;
use differential_dataflow::operators::consolidate::Consolidate;
use differential_dataflow::lattice::Lattice;
use timely::dataflow::scopes::Scope;
use timely::order::TotalOrder;
use differential_dataflow::operators::ThresholdTotal;
use std::mem;

pub fn graph_SCC<S,V,E,N,EF,LF>(edges: &Collection<S,V,Weight>, _edges: EF,
                                from: fn(&E) -> N,
                                to:   fn(&E) -> N,
                                _scclabels: LF) -> (Collection<S,V,Weight>)
where
     S: Scope,
     S::Timestamp: Lattice + Ord,
     V: Val,
     N: Val,
     E: Val,
     EF: Fn(V) -> E + 'static,
     LF: Fn((N,N)) -> V + 'static
{
    let pairs = edges.map(move |v| {
        let e = _edges(v);
        (from(&e), to(&e))
    });

    /* Recursively trim nodes without incoming and outgoing edges */
    let trimmed = scc::trim(&scc::trim(&pairs).map_in_place(|x| mem::swap(&mut x.0, &mut x.1)))
                  .map_in_place(|x| mem::swap(&mut x.0, &mut x.1));
    /* Edges that form cycles */
    let cycles = scc::strongly_connected(&trimmed);
    /* Initially each node is labeled by its own id */
    let nodes  = cycles.map_in_place(|x| x.0 = x.1.clone()).consolidate();
    /* Propagate smallest ID within SCC */
    let scclabels = propagate::propagate(&cycles, &nodes);
    scclabels.map(_scclabels)
}

pub fn graph_ConnectedComponents<S,V,E,N,EF,LF>(edges: &Collection<S,V,Weight>, _edges: EF,
                                                from: fn(&E) -> N,
                                                to:   fn(&E) -> N,
                                                _cclabels: LF) -> (Collection<S,V,Weight>)
where
     S: Scope,
     S::Timestamp: Lattice + Ord,
     V: Val,
     N: Val,
     E: Val,
     EF: Fn(V) -> E + 'static,
     LF: Fn((N,N)) -> V + 'static
{
    let pairs = edges.map(move |v| {
        let e = _edges(v);
        (from(&e), to(&e))
    });

    /* Initially each node is labeled by its own id */
    let nodes  = pairs.map_in_place(|x| x.0 = x.1.clone()).consolidate();
    let labels = propagate::propagate(&pairs, &nodes);
    labels.map(_cclabels)
}

pub fn graph_ConnectedComponents64<S,V,E,N,EF,LF>(edges: &Collection<S,V,Weight>, _edges: EF,
                                                  from: fn(&E) -> N,
                                                  to:   fn(&E) -> N,
                                                  _cclabels: LF) -> (Collection<S,V,Weight>)
where
     S: Scope,
     S::Timestamp: Lattice + Ord,
     u64: From<N>,
     V: Val,
     N: Val + Clone,
     E: Val,
     EF: Fn(V) -> E + 'static,
     LF: Fn((N,N)) -> V + 'static
{
    let pairs = edges.map(move |v| {
        let e = _edges(v);
        (from(&e), to(&e))
    });

    /* Initially each node is labeled by its own id */
    let nodes  = pairs.map_in_place(|x| x.0 = x.1.clone()).consolidate();

    /* `propagate_at` is the same as `propagate` but schedules the work by first circulating
     * all elements with the same value of the closure.
     * We use a closure that maps small numbers to small logarithms
     * to drop the footprint for the first iterative computation.
     */
    let labels = propagate::propagate_at(&pairs, &nodes, |x| u64::from(x.clone()));
    labels.map(_cclabels)
}

pub fn graph_UnsafeBidirectionalEdges<S,V,E,N,EF,LF>(edges: &Collection<S,V,Weight>, _edges: EF,
                                                     from: fn(&E) -> N,
                                                     to:   fn(&E) -> N,
                                                     _biedges: LF) -> (Collection<S,V,Weight>)
where
     S: Scope,
     S::Timestamp: TotalOrder + Lattice + Ord,
     V: Val,
     N: Val,
     E: Val,
     EF: Fn(V) -> E + 'static,
     LF: Fn((N,N)) -> V + 'static
{
    let mins = edges.map(move |v| {
        let e = _edges(v);
        let x = from(&e);
        let y = to(&e);
        if x < y { (x,y) } else { (y,x) }
    });

    let bidirectional = mins.threshold_total(|_, count| if *count > 1 { 1 } else { 0 });
    let bidirectional = bidirectional.concat(&bidirectional.map(|(x,y)|(y.clone(),x.clone())));
    bidirectional.map(_biedges)
}
