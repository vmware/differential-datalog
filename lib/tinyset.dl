/* `tinyset` (https://docs.rs/tinyset/0.2.2/tinyset/) is a Rust crate that
 * implements several space-efficient sparse collections.  This library is a
 * wrapper around `tinyset` that so far only supports one of these types --
 * `Set64`, which is a set of values that fit in 64 bits and can be cloned by
 * copying (e.g., 8, 16, 32, or 64-bit integers). */

extern type Set64<'X>

extern function size(s: Set64<'X>): bit<64>
extern function empty(): Set64<'X>
extern function singleton(x: 'X): Set64<'X>
extern function insert(s: mut Set64<'X>, v: 'X): ()
extern function insert_imm(s: Set64<'X>, v: 'X): Set64<'X>
extern function contains(s: Set64<'X>, v: 'X): bool
extern function is_empty(s: Set64<'X>): bool
extern function nth(s: Set64<'X>, n: bit<64>): Option<bit<64>>
extern function union(s1: Set64<'X>, s2: Set64<'X>): Set64<'X>
extern function unions(sets: Vec<Set64<'X>>): Set64<'X>
extern function intersection(s1: Set64<'X>, s2: Set64<'X>): Set64<'X>
extern function difference(s1: Set64<'X>, s2: Set64<'X>): Set64<'X>

/* Aggregates */
extern function group2set(g: Group<'K, 'V>): Set64<'V>
extern function group_set_unions(g: Group<'K, Set64<'V>>): Set64<'V>

/* Optimized version of group_set_unions that, when the group consits of
 * a single set, simply returns the reference to this set. */
extern function group_setref_unions(g: Group<'K, Ref<Set64<'V>>>): Ref<Set64<'V>>
