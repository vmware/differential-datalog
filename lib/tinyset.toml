[dependencies.tinyset]
version = "0.2.2"
