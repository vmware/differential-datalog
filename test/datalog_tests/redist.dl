import tinyset as TS
import graph as graph

/* entity id */
typedef entid_t = bit<32>

/* transport node id */
typedef tnid_t = bit<16>

input relation DdlogNode(id: entid_t)
input relation DdlogBinding(tn: tnid_t, entity: entid_t)
input relation DdlogDependency(parent: entid_t, child: entid_t)

/* Bi-directional edges.
 */
relation BiEdge(parent: entid_t, child: entid_t)
BiEdge(parent, child) :- DdlogDependency(parent, child), DdlogDependency(child, parent).

/* Compute connected components by propagating node id's along bidirectional edges.
 */
function edge_parent(e: BiEdge): entid_t = e.parent
function edge_child(e: BiEdge): entid_t = e.child
relation SCCLabel[(entid_t, entid_t)]
apply graph::ConnectedComponents(BiEdge, edge_child, edge_parent) -> (SCCLabel)

/* Label all nodes that don't belong to an SCC with their own ids
 */
relation LabeledNode(node: entid_t, scc: entid_t)
LabeledNode(node, scc) :- SCCLabel[(node, scc)].
LabeledNode(node, node) :- DdlogNode(node), not SCCLabel[(node, _)].

/* Lift the DdlogDependency relation to SCCs:
 * Two SCCs are connected by an edge iff there is there exists a pair
 * of nodes that belong to the first and the second SCC respectively,
 * connected by a DdlogDependency edge.
 */
relation SCCEdge(parent: entid_t, child: entid_t)
SCCEdge(parentscc,childscc) :-
    DdlogDependency(parent, child),
    LabeledNode(parent, parentscc),
    LabeledNode(child, childscc),
    childscc != parentscc.

/* Lift the DdlogBinding relation to SCCs:
 * An SCC is bound to a TN iff at least one of its nodes is bound to this TN.
 */
relation SCCBinding(scc: entid_t, bindings: Ref<TS::Set64<tnid_t>>)
SCCBinding(scc, ref_new(bindings)) :-
    DdlogBinding(tn, entity),
    LabeledNode(entity, scc),
    var bindings = Aggregate((scc), TS::group_to_set(tn)).

/* Compute SCC span.
 */
relation SCCSpan(scc: entid_t, span: Ref<TS::Set64<tnid_t>>)
/* Base case */
SCCSpan(scc, bindings) :- SCCBinding(scc, bindings).
/* Recursive step: propagate span along graph edges */
SCCSpan(parent, tns) :-
    SCCEdge(child, parent),
    SCCSpan(child, child_tns),
    var tns = Aggregate((parent), TS::group_setref_unions(child_tns)).

/* Flatten the span relation: for each node in an SCC, the node's span is the
 * same as the SCC span.
 */
output relation Span(entity: entid_t, tns: Ref<TS::Set64<tnid_t>>)
Span(node, span) :- SCCSpan(scc, span), LabeledNode(node, scc).

/* The following is validation code to check that the span computation is equivalent to a simple
 * two-line reachability computation.
 */
/*
output relation FlatSpan(entity: entid_t, tn: tnid_t)
FlatSpan(entity, tn) :-
    Span(entity, &span),
    var tn = FlatMap(span).

relation SimpleSpan(entity: entid_t, tn: tnid_t)
SimpleSpan(entity, tn) :- DdlogBinding(tn, entity).
SimpleSpan(parent, tn) :- DdlogDependency(child, parent), SimpleSpan(child, tn).

output relation Error(entity: entid_t, tn: tnid_t)
Error(entity, tn) :-
    FlatSpan(entity, tn),
    not SimpleSpan(entity, tn).

Error(entity, tn) :-
    SimpleSpan(entity, tn),
    not FlatSpan(entity, tn).
*/
