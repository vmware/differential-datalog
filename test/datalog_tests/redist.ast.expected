typedef BiEdge = BiEdge{parent: entid_t, child: entid_t}
typedef DdlogBinding = DdlogBinding{tn: tnid_t, entity: entid_t}
typedef DdlogDependency = DdlogDependency{parent: entid_t, child: entid_t}
typedef DdlogNode = DdlogNode{id: entid_t}
typedef LabeledNode = LabeledNode{node: entid_t, scc: entid_t}
typedef SCCBinding = SCCBinding{scc: entid_t, bindings: std.Ref<tinyset.Set64<tnid_t>>}
typedef SCCEdge = SCCEdge{parent: entid_t, child: entid_t}
typedef SCCSpan = SCCSpan{scc: entid_t, span: std.Ref<tinyset.Set64<tnid_t>>}
typedef Span = Span{entity: entid_t, tns: std.Ref<tinyset.Set64<tnid_t>>}
typedef entid_t = bit<32>
typedef log.log_level_t = signed<32>
typedef log.module_t = signed<32>
typedef std.Either<'A,'B> = std.Left{l: 'A} | std.Right{r: 'B}
extern type std.Group<'A>
extern type std.Map<'K,'V>
typedef std.Option<'A> = std.Some{x: 'A} | std.None{}
#[size = 8]
extern type std.Ref<'A>
extern type std.Set<'A>
extern type std.Vec<'A>
extern type tinyset.Set64<'X>
typedef tnid_t = bit<16>
function edge_child (e: BiEdge): entid_t =
    e.child
function edge_parent (e: BiEdge): entid_t =
    e.parent
extern function log.log (module: log.module_t, level: log.log_level_t, msg: string): bool
extern function std.__builtin_2string (x: 'X): string
extern function std.deref (x: std.Ref<'A>): 'A
extern function std.group2map (g: std.Group<('K, 'V)>): std.Map<'K,'V>
extern function std.group2set (g: std.Group<'A>): std.Set<'A>
extern function std.group2vec (g: std.Group<'A>): std.Vec<'A>
extern function std.group_count (g: std.Group<'A>): bit<64>
extern function std.group_first (g: std.Group<'A>): 'A
extern function std.group_max (g: std.Group<'A>): 'A
extern function std.group_min (g: std.Group<'A>): 'A
extern function std.group_nth (g: std.Group<'A>, n: bit<64>): std.Option<'A>
extern function std.group_set_unions (g: std.Group<std.Set<'A>>): std.Set<'A>
extern function std.group_setref_unions (g: std.Group<std.Ref<std.Set<'A>>>): std.Ref<std.Set<'A>>
extern function std.group_sum (g: std.Group<'A>): 'A
function std.group_unzip (g: std.Group<('X, 'Y)>): (std.Vec<'X>, std.Vec<'Y>) =
    ((var xs: std.Vec<'X>) = std.vec_empty();
     ((var ys: std.Vec<'Y>) = std.vec_empty();
      (for (v in g) {
           ((var x, var y) = v;
            (std.vec_push(xs, x);
             std.vec_push(ys, y)))
       };
       (xs, ys))))
extern function std.hash128 (x: 'X): bit<128>
extern function std.hash64 (x: 'X): bit<64>
extern function std.hex (x: 'X): string
extern function std.htonl (x: bit<32>): bit<32>
extern function std.htons (x: bit<16>): bit<16>
function std.is_none (x: std.Option<'A>): bool =
    match (x) {
        std.None{} -> true,
        _ -> false
    }
function std.is_some (x: std.Option<'A>): bool =
    match (x) {
        std.Some{.x=_} -> true,
        _ -> false
    }
extern function std.map_contains_key (m: std.Map<'K,'V>, k: 'K): bool
extern function std.map_empty (): std.Map<'K,'V>
extern function std.map_get (m: std.Map<'K,'V>, k: 'K): std.Option<'V>
extern function std.map_insert (m: mut std.Map<'K,'V>, k: 'K, v: 'V): ()
extern function std.map_insert_imm (m: std.Map<'K,'V>, k: 'K, v: 'V): std.Map<'K,'V>
extern function std.map_is_empty (m: std.Map<'K,'V>): bool
extern function std.map_remove (m: mut std.Map<'K,'V>, k: 'K): ()
extern function std.map_singleton (k: 'K, v: 'V): std.Map<'K,'V>
extern function std.map_union (m1: std.Map<'K,'V>, m2: std.Map<'K,'V>): std.Map<'K,'V>
extern function std.max (x: 'A, y: 'A): 'A
extern function std.min (x: 'A, y: 'A): 'A
extern function std.ntohl (x: bit<32>): bit<32>
extern function std.ntohs (x: bit<16>): bit<16>
extern function std.parse_dec_i64 (s: string): std.Option<signed<64>>
extern function std.parse_dec_u64 (s: string): std.Option<bit<64>>
extern function std.pow32 (base: 'A, exp: bit<32>): 'A
extern function std.range (from: 'A, to: 'A, step: 'A): std.Vec<'A>
extern function std.ref_new (x: 'A): std.Ref<'A>
extern function std.set2vec (s: std.Set<'A>): std.Vec<'A>
extern function std.set_contains (s: std.Set<'X>, v: 'X): bool
extern function std.set_empty (): std.Set<'X>
extern function std.set_insert (s: mut std.Set<'X>, v: 'X): ()
extern function std.set_insert_imm (s: std.Set<'X>, v: 'X): std.Set<'X>
extern function std.set_is_empty (s: std.Set<'X>): bool
extern function std.set_nth (s: std.Set<'X>, n: bit<64>): std.Option<'X>
extern function std.set_singleton (x: 'X): std.Set<'X>
extern function std.set_size (s: std.Set<'X>): bit<64>
extern function std.set_union (s1: std.Set<'X>, s2: std.Set<'X>): std.Set<'X>
extern function std.set_unions (sets: std.Vec<std.Set<'X>>): std.Set<'X>
extern function std.str_to_lower (s: string): string
extern function std.string_contains (s1: string, s2: string): bool
extern function std.string_join (strings: std.Vec<string>, sep: string): string
extern function std.string_len (s: string): bit<64>
extern function std.string_split (s: string, sep: string): std.Vec<string>
extern function std.string_substr (s: string, start: bit<64>, end: bit<64>): string
extern function std.vec2set (s: std.Vec<'A>): std.Set<'A>
extern function std.vec_contains (v: std.Vec<'X>, x: 'X): bool
extern function std.vec_empty (): std.Vec<'A>
extern function std.vec_is_empty (v: std.Vec<'X>): bool
extern function std.vec_len (v: std.Vec<'X>): bit<64>
extern function std.vec_nth (v: std.Vec<'X>, n: bit<64>): std.Option<'X>
extern function std.vec_push (v: mut std.Vec<'X>, x: 'X): ()
extern function std.vec_push_imm (v: std.Vec<'X>, x: 'X): std.Vec<'X>
extern function std.vec_singleton (x: 'X): std.Vec<'X>
extern function tinyset.contains (s: tinyset.Set64<'X>, v: 'X): bool
extern function tinyset.empty (): tinyset.Set64<'X>
extern function tinyset.group2set (g: std.Group<'A>): tinyset.Set64<'A>
extern function tinyset.group_set_unions (g: std.Group<tinyset.Set64<'A>>): tinyset.Set64<'A>
extern function tinyset.group_setref_unions (g: std.Group<std.Ref<tinyset.Set64<'A>>>): std.Ref<tinyset.Set64<'A>>
extern function tinyset.insert (s: mut tinyset.Set64<'X>, v: 'X): ()
extern function tinyset.insert_imm (s: tinyset.Set64<'X>, v: 'X): tinyset.Set64<'X>
extern function tinyset.is_empty (s: tinyset.Set64<'X>): bool
extern function tinyset.nth (s: tinyset.Set64<'X>, n: bit<64>): std.Option<bit<64>>
extern function tinyset.singleton (x: 'X): tinyset.Set64<'X>
extern function tinyset.size (s: tinyset.Set64<'X>): bit<64>
extern function tinyset.union (s1: tinyset.Set64<'X>, s2: tinyset.Set64<'X>): tinyset.Set64<'X>
extern function tinyset.unions (sets: std.Vec<tinyset.Set64<'X>>): tinyset.Set64<'X>
extern transformer graph.ConnectedComponents(Edges: relation['E], from: function(e: 'E): 'N, to: function(e: 'E): 'N) -> (CCLabels: relation[('N, 'N)])
extern transformer graph.ConnectedComponents64(Edges: relation['E], from: function(e: 'E): 'N, to: function(e: 'E): 'N) -> (CCLabels: relation[('N, 'N)])
extern transformer graph.SCC(Edges: relation['E], from: function(e: 'E): 'N, to: function(e: 'E): 'N) -> (SCCLabels: relation[('N, 'N)])
extern transformer graph.UnsafeBidirectionalEdges(Edges: relation['E], from: function(e: 'E): 'N, to: function(e: 'E): 'N) -> (BiEdges: relation[('N, 'N)])
relation BiEdge [BiEdge]
input relation DdlogBinding [DdlogBinding]
input relation DdlogDependency [DdlogDependency]
input relation DdlogNode [DdlogNode]
relation LabeledNode [LabeledNode]
relation SCCBinding [SCCBinding]
relation SCCEdge [SCCEdge]
relation SCCLabel [(entid_t, entid_t)]
relation SCCSpan [SCCSpan]
output relation Span [Span]
BiEdge(.parent=parent, .child=child) :- DdlogDependency(.parent=parent, .child=child), DdlogDependency(.parent=child, .child=parent).
LabeledNode(.node=node, .scc=scc) :- SCCLabel[(node, scc)].
LabeledNode(.node=node, .scc=node) :- DdlogNode(.id=node), not SCCLabel[(node, _)].
SCCEdge(.parent=parentscc, .child=childscc) :- DdlogDependency(.parent=parent, .child=child), LabeledNode(.node=parent, .scc=parentscc), LabeledNode(.node=child, .scc=childscc), (childscc != parentscc).
SCCBinding(.scc=scc, .bindings=std.ref_new(bindings)) :- DdlogBinding(.tn=tn, .entity=entity), LabeledNode(.node=entity, .scc=scc), var bindings = Aggregate((scc), tinyset.group2set(tn)).
SCCSpan(.scc=scc, .span=bindings) :- SCCBinding(.scc=scc, .bindings=bindings).
SCCSpan(.scc=parent, .span=tns) :- SCCEdge(.parent=child, .child=parent), SCCSpan(.scc=child, .span=child_tns), var tns = Aggregate((parent), tinyset.group_setref_unions(child_tns)).
Span(.entity=node, .tns=span) :- SCCSpan(.scc=scc, .span=span), LabeledNode(.node=node, .scc=scc).
apply graph.ConnectedComponents(BiEdge, edge_child, edge_parent) -> (SCCLabel)
